containers-tests/tests/map-properties.hs

```diff
@@ -1272,31 +1278,46 @@ instance Arbitrary WhenMatchedSpec where
 
 ----------------------------------------------------------------
 
-prop_ordered :: Property
-prop_ordered
-  = forAll (choose (5,100)) $ \n ->
-    let xs = [(x,()) | x <- [0..n::Int]]
-    in fromAscList xs == fromList xs
-
-prop_rev_ordered :: Property
-prop_rev_ordered
-  = forAll (choose (5,100)) $ \n ->
-    let xs = [(x,()) | x <- [0..n::Int]]
-    in fromDescList (reverse xs) == fromList xs
-
 prop_list :: [Int] -> Bool
 prop_list xs = (sort (nub xs) == [x | (x,()) <- toList (fromList [(x,()) | x <- xs])])
 
 prop_descList :: [Int] -> Bool
 prop_descList xs = (reverse (sort (nub xs)) == [x | (x,()) <- toDescList (fromList [(x,()) | x <- xs])])
 
+prop_fromDescList :: [(Int, A)] -> Property
+prop_fromDescList kxs =
+    valid t .&&.
+    t === fromList kxs
+  where
+    downSortedKxs = List.sortBy (comparing (Down . fst)) kxs
+    t = fromDescList downSortedKxs
+
+prop_fromDescListWith :: Fun (A, A) A -> [(Int, A)] -> Property
+prop_fromDescListWith f kxs =
+    valid t .&&.
+    t === fromListWith (apply2 f) downSortedKxs
+  where
+    downSortedKxs = List.sortBy (comparing (Down . fst)) kxs
+    t = fromDescListWith (apply2 f) downSortedKxs
+
+prop_fromDescListWithKey :: Fun (Int, A, A) A -> [(Int, A)] -> Property
+prop_fromDescListWithKey f kxs =
+    valid t .&&.
+    t === fromListWithKey (apply3 f) downSortedKxs
+  where
+    downSortedKxs = List.sortBy (comparing (Down . fst)) kxs
+    t = fromDescListWithKey (apply3 f) downSortedKxs
+
 prop_fromDistinctDescList :: [(Int, A)] -> Property
-prop_fromDistinctDescList xs =
+prop_fromDistinctDescList kxs =
     valid t .&&.
-    toList t === nub_sort_xs
+    toList t === reverse nubDownSortedKxs
   where
-    t = fromDistinctDescList (reverse nub_sort_xs)
-    nub_sort_xs = List.map List.head $ List.groupBy ((==) `on` fst) $ List.sortBy (comparing fst) xs
+    nubDownSortedKxs =
+      List.map NE.head $
```

Comment:
I know you didn't do it, but `map head q` is bletcherous. Why not `[x | x : _ <- q]` instead?

Reply:
What's wrong with that? It's `NonEmpty`'s `head` btw, not list.

Reply:
Ah... sorry ....